haystack/schema.py

```diff
@@ -16,21 +15,15 @@
 import time
 import json
 import ast
-from dataclasses import asdict
+from dataclasses import asdict, InitVar
```

Comment:
Given that you're not using `__post_init__` now, `InitVar` should not be necessary anymore, right? I'll mark the spots where this is used so you can remove it easily if it should be done.

```suggestion
from dataclasses import asdict
```

https://pydantic-docs.helpmanual.io/usage/dataclasses/#initialize-hooks

Reply:
What happens is a divergence between 3.7 and 3.9 regarding our custom initialization and when validation is fired. If we remove the InitVar and let the field on the dataclass, it will assume it's a defacto field, and not only one field used for initialization. Result is, it will validate and propagate it. Even for generated schemas and models. Which I think It should not be.

The declaration as InitVar is what it should be, a dataclass with some parameter sent just on initialization. The standard. But, the __post_init__ are being mishandled on 3.7.

I really think we should let it, and in the future refactor the codebase starting from there.